test/test_linearizer.py

```diff
@@ -79,6 +80,42 @@ def test_load_removed(self):
     np.testing.assert_equal(a.numpy(), ta)
     np.testing.assert_equal(b.numpy(), tb)
 
+  def test_multireduce(self):
+    x = Tensor.empty(32)
+    r0 = x.max(axis=0, keepdim=True)
+    r1 = (x - r0).sum(axis=0)
+    out = r1 + 2
+    ast = create_schedule([out.lazydata])[-1].ast
+    for op in ast: print_tree(op)
+    lin = Linearizer(*ast)
+    assert lin.reduceops[0].op is ReduceOps.MAX
+    assert lin.reduceops[1].op is ReduceOps.SUM
+
+    x = Tensor.empty(32, 32)
+    y = Tensor.empty(1, 32)
+    r0 = x.max(axis=0, keepdim=True)
+    out0 = r0 + y
+    out1 = r0 * y
+    r1 = (x - out0).sum(axis=0)
+    out2 = r1 + 10
+    ast = create_schedule([out2.lazydata, out0.lazydata, out1.lazydata])[-1].ast
+    for op in ast: print_tree(op)
+    lin = Linearizer(*ast)
+    assert lin.reduceops[0].op is ReduceOps.MAX
+    assert lin.reduceops[1].op is ReduceOps.SUM
+
+    x = Tensor.empty(32, 32)
+    y = Tensor.empty(1, 32)
+    r0 = x.max(axis=0, keepdim=True)
+    r1 = (x - r0).sum(axis=0)
+    out = r0 + r1
+    ast = create_schedule([out.lazydata])[-1].ast
+    for op in ast: print_tree(op)
+    lin = Linearizer(*ast)
+    assert lin.reduceops[0].op is ReduceOps.MAX
```

Comment:
@0xtimmy this assert should pass.

Reply:
oh good catch, I think this can be solved by just switching around where `[self]` is added in ops.py:
`def lazyops(self) -> List[LazyOp]: return dedup([self] + [item for x in self.src for item in x.lazyops])`
becomes:
`def lazyops(self) -> List[LazyOp]: return dedup([item for x in self.src for item in x.lazyops] + [self])`

so that deeper lazyops always come first

or if we don't want to change the LazyOp code we could put a helper function in `kernel.py` like this:
```
ordered_lazyop_dict = {}
  def ordered_lazyops(op): 
    if op not in ordered_lazyop_dict: ordered_lazyop_dict[op] = dedup([item for x in op.src for item in ordered_lazyops(x)] + [op])
    return ordered_lazyop_dict[op]
  self.reduceops = dedup([x for out in self.ast for x in ordered_lazyops(out) if x.op in ReduceOps])
```